File: FNM_RE/shared/text.py
"""FNM_RE 文本相关工具。"""

from __future__ import annotations

import re
from typing import Any, Mapping

from FNM_RE.shared.title import normalize_title, normalized_title_key

_MARKDOWN_HEADING_RE = re.compile(r"^\s{0,3}#{1,6}\s*(.+?)\s*$")
_NOTES_HEADER_RE = re.compile(r"^\s*(?:#+\s*)?(notes?|endnotes?|notes to pages?.*)\s*$", re.IGNORECASE)


def page_markdown_text(page: Mapping[str, Any] | None) -> str:
    if not isinstance(page, Mapping):
        return ""
    markdown = page.get("markdown")
    if isinstance(markdown, Mapping):
        return str(markdown.get("text") or "").strip()
    if markdown:
        return str(markdown).strip()
    nested_page = page.get("_page")
    if isinstance(nested_page, Mapping) and nested_page is not page:
        return page_markdown_text(nested_page)
    return ""


def page_blocks(page: Mapping[str, Any] | None) -> list[dict[str, Any]]:
    payload = dict(page or {})
    pruned = dict(payload.get("prunedResult") or {})
    blocks = list(pruned.get("parsing_res_list") or [])
    return sorted(
        [dict(block) for block in blocks if isinstance(block, dict)],
        key=lambda block: (
            int(block.get("block_order", 10**9) or 10**9),
            float((block.get("block_bbox") or [0, 0, 0, 0])[1] or 0),
            float((block.get("block_bbox") or [0, 0, 0, 0])[0] or 0),
        ),
    )


def extract_page_headings(page: Mapping[str, Any] | None) -> list[str]:
    headings: list[str] = []
    seen: set[str] = set()
    for block in page_blocks(page):
        label = str(block.get("block_label") or "").strip().lower()
        if label not in {"doc_title", "paragraph_title"}:
            continue
        text = normalize_title(str(block.get("block_content") or ""))
        key = normalized_title_key(text)
        if text and key and key not in seen:
            seen.add(key)
            headings.append(text)
    if headings:
        return headings
    for raw_line in page_markdown_text(page).splitlines()[:8]:
        match = _MARKDOWN_HEADING_RE.match(raw_line)
        if not match:
            continue
        text = normalize_title(match.group(1))
        key = normalized_title_key(text)
        if text and key and key not in seen:
            seen.add(key)
            headings.append(text)
    return headings


def has_note_heading(page: Mapping[str, Any] | None) -> bool:
    return bool(_NOTES_HEADER_RE.search(page_markdown_text(page)))


def first_section_hint(page: Mapping[str, Any] | None, note_scan: Mapping[str, Any] | None) -> str:
    headings = extract_page_headings(page)
    if headings:
        return headings[0]
    hints = list((note_scan or {}).get("section_hints") or [])
    if hints:
        return str(hints[0] or "")
    return ""


def note_scan_summary(note_scan: Mapping[str, Any] | None) -> dict[str, Any]:
    scan = dict(note_scan or {})
    return {
        "page_kind": scan.get("page_kind"),
        "item_count": len(scan.get("items") or []),
        "ambiguity_flags": list(scan.get("ambiguity_flags") or []),
        "note_start_line_index": scan.get("note_start_line_index"),
    }


def plain_text_lines(text: str) -> list[str]:
    return [line.strip() for line in str(text or "").splitlines() if str(line or "").strip()]

File: fnm_re_rs/fnm-core/src/text.rs
//! ←→ FNM_RE/shared/text.py
//! 页面文本提取与 heading 检测工具。
//!
//! 输入 page 使用 `&serde_json::Value` 保持 Python dict 兼容。

use crate::title::{normalize_title, normalized_title_key};
use once_cell::sync::Lazy;
use regex::Regex;
use serde_json::Value;

static MARKDOWN_HEADING_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^\s{0,3}#{1,6}\s*(.+?)\s*$").unwrap());

static NOTES_HEADER_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?im)^\s*(?:#+\s*)?(?:notes?|endnotes?|notes to pages?.*|注释|脚注|尾注)\s*$")
        .unwrap()
});

/// 从 page 对象提取 markdown 文本。与 Python `page_markdown_text` 一致。
pub fn page_markdown_text(page: &Value) -> String {
    if !page.is_object() {
        return String::new();
    }
    if let Some(md) = page
        .get("enriched_markdown")
        .or_else(|| page.get("markdown"))
    {
        if md.is_object() {
            return md
                .get("text")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .trim()
                .to_string();
        }
        if let Some(s) = md.as_str() {
            return s.trim().to_string();
        }
    }
    if let Some(nested) = page.get("_page") {
        if nested.is_object() && nested.as_object() != page.as_object() {
            return page_markdown_text(nested);
        }
    }
    String::new()
}

/// 从 page 的 prunedResult 提取排序后的 block 列表。
/// 与 Python `page_blocks` 一致。
pub fn page_blocks(page: &Value) -> Vec<Value> {
    let blocks = page
        .get("prunedResult")
        .and_then(|p| p.get("parsing_res_list"))
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();

    if blocks.is_empty() {
        return vec![];
    }

    let mut blocks: Vec<Value> = blocks.into_iter().filter(|b| b.is_object()).collect();

    blocks.sort_by(|a, b| {
        let order_a = a
            .get("block_order")
            .and_then(|v| v.as_i64())
            .unwrap_or(i64::MAX);
        let order_b = b
            .get("block_order")
            .and_then(|v| v.as_i64())
            .unwrap_or(i64::MAX);
        if order_a != order_b {
            return order_a.cmp(&order_b);
        }
        let get_bbox_1 = |v: &Value| -> f64 {
            v.get("block_bbox")
                .and_then(|b| b.as_array())
                .and_then(|arr| arr.get(1))
                .and_then(|v| v.as_f64())
                .unwrap_or(0.0)
        };
        let y_a = get_bbox_1(a);
        let y_b = get_bbox_1(b);
        if y_a != y_b {
            return y_a.partial_cmp(&y_b).unwrap_or(std::cmp::Ordering::Equal);
        }
        let get_bbox_0 = |v: &Value| -> f64 {
            v.get("block_bbox")
                .and_then(|b| b.as_array())
                .and_then(|arr| arr.first())
                .and_then(|v| v.as_f64())
                .unwrap_or(0.0)
        };
        let x_a = get_bbox_0(a);
        let x_b = get_bbox_0(b);
        x_a.partial_cmp(&x_b).unwrap_or(std::cmp::Ordering::Equal)
    });

    blocks
}

/// 从 page 提取标题行（去重）。与 Python `extract_page_headings` 一致。
pub fn extract_page_headings(page: &Value) -> Vec<String> {
    let mut headings: Vec<String> = Vec::new();
    let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();

    for block in page_blocks(page) {
        let label = block
            .get("block_label")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_lowercase();
        if label != "doc_title" && label != "paragraph_title" {
            continue;
        }
        let text = normalize_title(
            block
                .get("block_content")
                .and_then(|v| v.as_str())
                .unwrap_or(""),
        );
        let key = normalized_title_key(&text);
        if !text.is_empty() && !key.is_empty() && !seen.contains(&key) {
            seen.insert(key);
            headings.push(text);
        }
    }
    if !headings.is_empty() {
        return headings;
    }
    let md = page_markdown_text(page);
    for line in md.lines().take(8) {
        if let Some(caps) = MARKDOWN_HEADING_RE.captures(line) {
            let text = normalize_title(caps.get(1).map(|m| m.as_str()).unwrap_or(""));
            let key = normalized_title_key(&text);
            if !text.is_empty() && !key.is_empty() && !seen.contains(&key) {
                seen.insert(key);
                headings.push(text);
            }
        }
    }
    headings
}

/// 判断 page 是否包含 note heading。与 Python `has_note_heading` 一致。
pub fn has_note_heading(page: &Value) -> bool {
    NOTES_HEADER_RE.is_match(&page_markdown_text(page))
}

/// 返回 page 的第一个 section hint（标题或 note_scan section_hints）。
/// 与 Python `first_section_hint` 一致。
pub fn first_section_hint(page: &Value, note_scan: Option<&Value>) -> String {
    let headings = extract_page_headings(page);
    if let Some(h) = headings.first() {
        return h.clone();
    }
    if let Some(scan) = note_scan {
        if let Some(hints) = scan.get("section_hints").and_then(|v| v.as_array()) {
            if let Some(hint) = hints.first().and_then(|v| v.as_str()) {
                return hint.to_string();
            }
        }
    }
    String::new()
}

/// 生成 note_scan 的简短摘要。与 Python `note_scan_summary` 一致。
pub fn note_scan_summary(note_scan: &Value) -> Value {
    let mut summary = serde_json::Map::new();
    if let Some(kind) = note_scan.get("page_kind") {
        summary.insert("page_kind".into(), kind.clone());
    }
    let item_count = note_scan
        .get("items")
        .and_then(|v| v.as_array())
        .map(|a| a.len())
        .unwrap_or(0);
    summary.insert("item_count".into(), Value::Number(item_count.into()));
    summary.insert(
        "ambiguity_flags".into(),
        note_scan
            .get("ambiguity_flags")
            .cloned()
            .unwrap_or(Value::Array(vec![])),
    );
    if let Some(idx) = note_scan.get("note_start_line_index") {
        summary.insert("note_start_line_index".into(), idx.clone());
    }
    Value::Object(summary)
}

/// 按行拆分并去除空行/空白行。与 Python `plain_text_lines` 一致。
pub fn plain_text_lines(text: &str) -> Vec<String> {
    text.lines()
        .map(|l| l.trim().to_string())
        .filter(|l| !l.is_empty())
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn markdown_text_from_enriched() {
        let page = json!({"enriched_markdown": {"text": "## Notes\ncontent"}});
        assert_eq!(page_markdown_text(&page), "## Notes\ncontent");
    }

    #[test]
    fn markdown_text_from_raw() {
        let page = json!({"markdown": "plain text"});
        assert_eq!(page_markdown_text(&page), "plain text");
    }

    #[test]
    fn markdown_text_nested() {
        let page = json!({"_page": {"markdown": "nested text"}});
        assert_eq!(page_markdown_text(&page), "nested text");
    }

    #[test]
    fn markdown_text_empty() {
        assert_eq!(page_markdown_text(&json!(null)), "");
        assert_eq!(page_markdown_text(&json!({})), "");
    }

    #[test]
    fn has_note_heading_detect() {
        let page = json!({"markdown": "## Notes\ncontent"});
        assert!(has_note_heading(&page));
    }

    #[test]
    fn has_note_heading_false() {
        let page = json!({"markdown": "## Chapter 1"});
        assert!(!has_note_heading(&page));
    }

    #[test]
    fn plain_lines_basic() {
        let lines = plain_text_lines("  a  \n\n  b  \n  ");
        assert_eq!(lines, vec!["a", "b"]);
    }
}
